### ISSUE TITLE
Access raw registration value

### ISSUE THREAD

**Issue by klausXR** [jeffijoe/awilix#278]
Given this code

```js
class MyClass {
    constructor() {}
}

container.register({
    myClass: awilix.asClass(MyClass).scoped()
})
```

Is there a way to access the `MyClass` reference ( not the resolved value, but the class reference )?

My use case is that I'm trying to build a small plugin system for my application, so that I can implement specific functionalities as separate modules.

Here's a sample API for an event plugin

```js
@Event("product.created")
class ProductCreatedListener {
    constructor({...}) {}

    async handle(product) {}
}

// ...

awilix.register({
    ProductCreatedListener: awilix.asClass(ProductCreatedListener).scoped()
})
```

Any module can register an event listener via the `@Event` decorator, and as long as its registered in the container, it will be subscribed to the event automatically. 

To get this working, I can, for example, make the `@Event` decorator add some metadata to the `ProductCreatedListener` class

```js
export const Event = (eventName) => target => {
  target[__EVENT_LISTENER__] = {...}
}
```

Then, when the Event plugin is initialized, it should scan all registrations from the container and get references of the classes that have the `__EVENT_LISTENER__` property

Something like this ( pseudo code )

```js
class EventPlugin {
    constructor({ container }) {
        this.container = container
    }

    onInit() {
        const eventListeners = Object.values(this.container.registrations)
            .filter(registration => {
                // property doesnt exist
                return registration.raw['__EVENT_LISTENER__']
            })
            .map(registration => registration.raw)
        
        // Remaining code to subscribe all listeners to the events
    }
}
```

**Comment by jeffijoe** [comment]
I'm afraid not, but you can create your own resolver that wraps the `asClass` resolver and use that when registering event handlers?

**Comment by klausXR** [comment]
Do you think this is a valuable feature to include in the core library? I assumed that I can hack around it and get it to work as expected, however, it creates tight coupling between my code and the library and I try to avoid to do that, whenever possible, because it makes version updates that much difficult, especially when this pattern is repeated in many places, lots of hidden bugs tend to occur.

**Comment by jeffijoe** [comment]
Since it would only be relevant for certain types of resolvers, it would be difficult to implement using the `buildResolver` which adds the various methods for configuring various shared properties such as lifetime, disposer, etc.

I tend to stay away from NestJS-style fancy magic. 😄  Like, is it _really_ that much work to add a `.addListener` for each new event handler that it justifies +20x the amount of code than the amount of `.addListener` lines would be needed? 😉

**Comment by klausXR** [comment]
I think this feature would be very useful, and not only for classes, I think it should be available for all resolvers.

> I tend to stay away from NestJS-style fancy magic. 😄 Like, is it really that much work to add a .addListener for each new event handler that it justifies +20x the amount of code than the amount of .addListener lines would be needed? 😉

I simply referred to it to showcase that its an actual use case and not a made up scenario.

About the second part of your question, yes, in some cases it might be a lot of work. This leaks implementation details to the business logic and transfers the actual wiring of a module to the one using it, not to the one that built it, which is backwards.

Its more declarative to let the user worry about his part in this ( which is, to which event he listens to ) as opposed to how exactly he should attach to it.

The declarative approach also allows you to do smarter clean ups and avoid memory leaks. If transfer to him the responsibility to `addListener` then he must also `.removeListener`. 

But this issue really repeats itself in many scenarios.

I don't have much experience with dependency containers, but I find it odd that the raw value is not exposed by default

**Comment by jeffijoe** [comment]
Because the following is a valid `Resolver`:

```js
container.register({
  greeting: {
    resolve: () => 'Hello world'
  }
})

container.resolve('greeting') // "Hello world"
```

A resolver does not necessarily have an underlying "raw" value.

> This leaks implementation details to the business logic and transfers the actual wiring of a module to the one using it, not to the one that built it, which is backwards.

A counterargument to this is that it makes it hard to follow the code in order to find out how something gets called/included. It's a trade-off for sure.

**Comment by klausXR** [comment]
What I meant by raw value is not the value after resolve is called, but the thing that will be used to get the value.

```js
class MyClass {}

container.register({
    greeting: {
        // In this example, the function should be the "raw" value, not the returned string
        // Because the function is the thing that gets called
        resolve: () => 'Hello World'
    },
    // In this example, the string "World!" should be the raw value, because nothing
    // is used to create it and it cannot change
    hello: awilix.asValue("World!"),

    // In this example, the MyClass constructor should be the raw value ( not an instance of it ),
    myClass: awilix.asClass(MyClass).scoped()
})
```

> A counterargument to this is that it makes it hard to follow the code in order to find out how something gets called/included. It's a trade-off for sure.

This is a trade off that is more than acceptable when authoring plugins or frameworks, but I really don't want to argue about it. I am well aware that there are trade offs that I can make to get around this, but I really believe this to be a worthy feature for a dependency container.

**Comment by jeffijoe** [comment]
> `// Because the function is the thing that gets called`

`resolve` won't be the value you are looking for though, because `asClass().resolve` won't be the class itself.

Due to the [immutable nature](https://github.com/jeffijoe/awilix/blob/e3cda5a153c0702c84f86478b2e22e0e9dca37d4/src/resolvers.ts?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L206) of how the internal resolvers work, I don't think this will be straightforward to implement in the container.

With that said, you can wrap any resolver to do what you want in userland code.
